readme.rst
.. default-role:: code

Disk Bench
##########

Designed to run similar tests as Crystal Disk Mark using fio with user friendly output options.

Tests ran:

* Sequential read/write w/ 1MB block size
* Random read/write w/ 512K block size
* Queue depth 32 random read/write 4K blocks size

Install
=======

::

    # system install
    $ sudo pip3 install disk-bench

    # user install
    $ pip3 install --user disk-bench

Usage
=====

::

    $ disk-bench --help

    # Default runs fio w/ --loops=3 and --size=1G
    $ disk-bench /mnt/disk-to-test/whatever

    # Show CSV output (for easy copy/paste into Excel or Google Sheets)
    $ disk-bench /mnt/disk-to-test/whatever --style=csv

    # Quick
    $ disk-bench /mnt/disk-to-test/whatever --loops=1 --size=1M


Command Details
===============

The `fio` command we are running is essentially:

    fio --directory . --filename=disk~bench.tmp --output-format=json --stonewall \
    --ioengine=libaio --direct=0 --gtod_reduce=1 \
    --name=seqread --bs=1m --rw=read --size=64G \
    --name=seqwrite --bs=1m --rw=write --size=64G \
    --name=randread --bs=512k --rw=randread --size=4G \
    --name=randwrite --bs=512k --rw=randwrite --size=4G \
    --name=4kQD32read --bs=4k --iodepth=32 --rw=randread --size=4G \
    --name=4kQD32write --bs=4k --iodepth=32 --rw=randwrite --size=4G \
    --name=4kQD16 --bs=4k --iodepth=32 --rw=randrw --rwmixread=65 --size=4G


changelog.rst
Changelog
=========

0.2.1 released 2018-12-22
-------------------------

- add fio command example to readme (d4f6bef_)
- add direct flag back (578f509_)

.. _d4f6bef: https://github.com/rsyring/disk-bench/commit/d4f6bef
.. _578f509: https://github.com/rsyring/disk-bench/commit/578f509


0.2.0 released 2018-12-21
-------------------------

- change tests we run and output format (df6ac7b_)

.. _df6ac7b: https://github.com/rsyring/disk-bench/commit/df6ac7b


0.1.4 released 2018-12-21
-------------------------

- fix cli direct flag (74aa304_)

.. _74aa304: https://github.com/rsyring/disk-bench/commit/74aa304


0.1.3 released 2018-12-21
-------------------------

- support non-direct IO (f5ba899_)

.. _f5ba899: https://github.com/rsyring/disk-bench/commit/f5ba899


0.1.2 released 2018-12-20
-------------------------

- fix some packaging issues (f908a1d_)

.. _f908a1d: https://github.com/rsyring/disk-bench/commit/f908a1d


0.1.1 released 2018-12-20
-------------------------

- fix pypi name (10d6115_)

.. _10d6115: https://github.com/rsyring/disk-bench/commit/10d6115

0.1.0 released 2018-12-20
-------------------------

- add tox & CI (12cfca5_)
- fix json bytes/text for Python 3.5 (475e3ee_)
- add tox improve package (ee8efc4_)

.. _12cfca5: https://github.com/rsyring/disk-bench/commit/12cfca5
.. _475e3ee: https://github.com/rsyring/disk-bench/commit/475e3ee
.. _ee8efc4: https://github.com/rsyring/disk-bench/commit/ee8efc4



diskbench/tests/test_cli.py
from click.testing import CliRunner

from diskbench import cli
from diskbench import fio
from diskbench import testing


def invoke_cli(*args):
    runner = CliRunner()
    result = runner.invoke(cli.db, args, catch_exceptions=False)
    assert result.exit_code == 0, result.output
    return result


def test_cli_csv(tmpdir):
    result = invoke_cli(tmpdir.strpath, '--style=csv', '--rand-size=512k', '--seq-size=1m')

    assert 'Stats (MB/s),seqread,randread' in result.output


def test_format_stats_csv():
    data = testing.load_json('fio-output.json')
    stats = fio.extract_stats(data)
    output = cli.format_stats(stats, 'csv')
    expected = '''
Stats (MB/s),seqread,randread,4kQD32read,4kQD16read,seqwrite,randwrite,4kQD32write,4kQD16write
,722.1,209.3,163.0,118.5,"2,909.1","2,709.0","1,321.3",63.7
'''.lstrip()
    assert output == expected


def test_format_stats_table():
    data = testing.load_json('fio-output.json')
    stats = fio.extract_stats(data)
    output = cli.format_stats(stats, 'table')
    # Read from file to avoid flake8 errors b/c table is pretty wide
    expected = testing.read('table-output.txt')
    assert output == expected


diskbench/tests/test_fio.py
import mock

from diskbench import fio
from diskbench.testing import load_json


def test_fio(tmpdir):
    stats = fio.fio(tmpdir.strpath, seq_size='1m', rand_size='512k', direct=False)

    job_stats = stats[0]
    assert job_stats.name == 'seqread'
    assert job_stats.bw

    job_stats = stats[1]
    assert job_stats.name == 'randread'
    assert job_stats.bw

    job_stats = stats[2]
    assert job_stats.name == '4kQD32read'
    assert job_stats.bw

    job_stats = stats[3]
    assert job_stats.name == '4kQD16read'
    assert job_stats.bw

    job_stats = stats[4]
    assert job_stats.name == 'seqwrite'
    assert job_stats.bw

    job_stats = stats[5]
    assert job_stats.name == 'randwrite'
    assert job_stats.bw

    job_stats = stats[6]
    assert job_stats.name == '4kQD32write'
    assert job_stats.bw

    job_stats = stats[7]
    assert job_stats.name == '4kQD16write'
    assert job_stats.bw


@mock.patch.object(fio, 'sh')
@mock.patch.object(fio, 'extract_stats')
def test_fio_direct(m_es, m_sh):
    m_sh.fio.return_value.stderr = None
    m_sh.fio.return_value.stdout = b'{}'

    fio.fio('/some/path', seq_size='1m', rand_size='512k', direct=True)
    args, kwargs = m_sh.fio.call_args
    assert args[6] == '--direct=1'

    fio.fio('/some/path', seq_size='1m', rand_size='512k', direct=False)
    args, kwargs = m_sh.fio.call_args
    assert args[6] == '--direct=0'


def test_stats_extract():
    data = load_json('fio-output.json')
    stats = fio.extract_stats(data)

    job_stats = stats[0]
    assert job_stats.name == 'seqread'
    assert job_stats.bw == 739475 / 1024

    job_stats = stats[1]
    assert job_stats.name == 'randread'
    assert job_stats.bw == 214345 / 1024

    job_stats = stats[2]
    assert job_stats.name == '4kQD32read'
    assert job_stats.bw == 166864 / 1024

    job_stats = stats[3]
    assert job_stats.name == '4kQD16read'
    assert job_stats.bw == 121329 / 1024

    job_stats = stats[4]
    assert job_stats.name == 'seqwrite'
    assert job_stats.bw == 2978909 / 1024

    job_stats = stats[5]
    assert job_stats.name == 'randwrite'
    assert job_stats.bw == 2774010 / 1024

    job_stats = stats[6]
    assert job_stats.name == '4kQD32write'
    assert job_stats.bw == 1353001 / 1024

    job_stats = stats[7]
    assert job_stats.name == '4kQD16write'
    assert job_stats.bw == 65216 / 1024


diskbench/cli.py
import csv
import io
import logging

import click
import tableprint as tp

from diskbench import fio

log = logging.getLogger(__name__)


@click.command()
@click.argument('fpath', type=click.Path(exists=False))
@click.option('--seq-size', default='64G')
@click.option('--rand-size', default='4G')
@click.option('--style', default='table', type=click.Choice(['table', 'csv']))
@click.option('--direct', is_flag=True, default=False)
def db(fpath, style, seq_size, rand_size, direct):
    stats = fio.fio(fpath, seq_size, rand_size, direct)
    click.echo(format_stats(stats, style))


def format_stats(stats, style):
    def format_stat(stat):
        return (
            stat.name,
            '{:,.1f}'.format(stat.bw),
            '{:,.0f}'.format(stat.iops),
            '{:,.1f}%'.format(stat.usr_cpu),
            '{:,.1f}%'.format(stat.sys_cpu),
        )

    out_fo = io.StringIO()
    header = ['Stats (MB/s)']
    header += [stat.name for stat in stats]
    table_rows = [[''] + ['{:,.1f}'.format(stat.bw) for stat in stats]]

    assert style in ('table', 'csv')
    if style == 'table':
        tp.table(table_rows, header, out=out_fo, style='round', width=12)
    else:
        writer = csv.writer(out_fo, lineterminator='\n')
        writer.writerow(header)
        writer.writerows(table_rows)

    out_fo.seek(0)
    return out_fo.read()


diskbench/fio.py
from collections import namedtuple
import os

import json

import sh

JobStat = namedtuple('JobStat', 'name bw iops usr_cpu sys_cpu')


def fio(dpath, seq_size, rand_size, direct):
    randsz_opt = '--size={}'.format(rand_size)
    seqsz_opt = '--size={}'.format(seq_size)
    direct_val = '1' if direct else '0'

    result = sh.fio(
        '--directory', dpath,
        '--filename=disk~bench.tmp',
        '--output-format=json',
        '--stonewall',
        '--ioengine=libaio',
        '--direct={}'.format(direct_val),
        '--gtod_reduce=1',
        '--name=seqread', '--bs=1m', '--rw=read', seqsz_opt,
        '--name=seqwrite', '--bs=1m', '--rw=write', seqsz_opt,
        '--name=randread', '--bs=512k', '--rw=randread', randsz_opt,
        '--name=randwrite', '--bs=512k', '--rw=randwrite', randsz_opt,
        '--name=4kQD32read', '--bs=4k', '--iodepth=32', '--rw=randread', randsz_opt,
        '--name=4kQD32write', '--bs=4k', '--iodepth=32', '--rw=randwrite', randsz_opt,
        '--name=4kQD16', '--bs=4k', '--iodepth=32', '--rw=randrw', '--rwmixread=65', randsz_opt,
    )

    db_tmp_fpath = os.path.join(dpath, 'disk~bench.tmp')
    if os.path.exists(db_tmp_fpath):
        os.unlink(db_tmp_fpath)

    assert not result.stderr, result.stderr

    fio_data = json.loads(result.stdout.decode('utf-8'))

    return extract_stats(fio_data)


def extract_stats(fio_data):
    return (
        extract_job_stats(fio_data['jobs'][0], 'read'),
        extract_job_stats(fio_data['jobs'][2], 'read'),
        extract_job_stats(fio_data['jobs'][4], 'read'),
        extract_job_stats(fio_data['jobs'][6], 'read', True),
        extract_job_stats(fio_data['jobs'][1], 'write'),
        extract_job_stats(fio_data['jobs'][3], 'write'),
        extract_job_stats(fio_data['jobs'][5], 'write'),
        extract_job_stats(fio_data['jobs'][6], 'write', True),
    )


def extract_job_stats(job_data, io_key, io_key_name=False):
    job_name = job_data['jobname']
    if io_key_name:
        job_name += io_key

    return JobStat(
        job_name,
        job_data[io_key]['bw'] / 1024,
        job_data[io_key]['iops'],
        job_data['usr_cpu'],
        job_data['sys_cpu'],
    )
